lib/collection/src/shards/queue_proxy_shard.rs

```diff
@@ -290,6 +330,26 @@ impl Inner {
         shard
     }
 
+    pub fn new_from_version(
+        wrapped_shard: LocalShard,
+        remote_shard: RemoteShard,
+        wal_keep_from: Arc<AtomicU64>,
+        version: u64,
+    ) -> Self {
+        let shard = Self {
+            wrapped_shard,
+            remote_shard,
+            last_update_idx: version.into(),
+            update_lock: Default::default(),
+            wal_keep_from,
+        };
+
+        // Keep all WAL entries from `version` so we don't truncate them off when we still need to transfer
+        shard.set_wal_keep_from(Some(version));
```

Comment:
In `Inner::new` we keep `last_update_idx` as-is, but `set_wal_keep_from(last_idx + 1)`. Can there be an "off by 1" error hidden here? E.g., should we do `last_update_idx = version - 1`?

(Note, this is a *question*, not a suggestion. Just noted it, cause `ClockSet`/`ClockMap` do have this distinction, maybe this case is similar?)

Reply:
I'd prefer to keep it this way. That is because both variables/parameters describe two different things:

- `last_update_idx` describes the last update we have actually transferred, which matches the last operation index
- `wal_keep_from` describes what version we want to keep

We can rename `last_update_idx` to `first_not_transferred_idx` or something similar, but I don't really like this. Also, we cannot change `wal_keep_from`.

Before, `wal_keep_from` was named `max_ack_version`, in which case the two matched up better.

> Also, we probably should rename `last_update_idx` to something more meaningful

What about `last_transferred_idx`?

Reply:
I still don't get why we do `+1` in one case, and we don't in another.

Like, my concern here, is that if there's some other block of logic (outside `new`/`new_from_version`; or maybe we will add one in the future) handling `last_update_idx` and making this assumption of "we keep WAL from `last_update_idx + 1`", we may have an annoying bug. And I'm not sure if we can't easily distinguish these two cases looking only at the `QueueProxy` fields.

Reply:
Though, if you sure that this is fine, then it's fine by me. At least we know where to look if we suddenly see first operation not being transferred or something. 😂

Reply:
There was a mistake indeed. In case of `new_from_version` the `last_update_idx` assignment was wrong.

I've fixed this and changed the field to `transfer_from` to make it work.